src/lib/optimizer/strategy/subquery_to_join_rule.cpp

```diff
@@ -0,0 +1,583 @@
+  // Check if node contains a subquery and turn it into an anti- or semi-join if possible.
+  // To do this, we
+  //   - Check whether node is of a supported type:
+  //       - (NOT) IN predicate with a subquery as the right operand
+  //       - (NOT) EXISTS predicate
+  //       - comparison (<,>,<=,>=,=,<>) predicate with subquery as the right operand
+  //   - If node is a (NOT) IN or a comparison, extract a first join predicate
+  //   - Scan the LQP for all usages of correlated parameters, counting the number of predicate nodes using them (if
+  //     one is used outside of predicate nodes, we never optimize the LQP).
+  //   - Recursively scan for correlated predicate nodes that we can safely remove (pull up and out of the subquery).
+  //     If we find all correlated predicate nodes, we copy and adapt the LQP while unwinding from the recursion.
+  //     Changes to non-predicate nodes are necessary to have all columns required by the new join predicates available
+  //     at the top of the subqueries LQP.
+  //   - Remove found predicates and adjust the subqueries LQP to allow them to be re-inserted as join predicates
+  //     (remove projections pruning necessary columns, etc.)
+  //   - Build a join with the collected predicates
+  //
+  // We always reformulate when possible, since benchmarks have shown that this reformulation makes the execution
+  // faster regardless of the expected table sizes, etc.
+
+  // Check whether the input LQP is supported, and extract some unified information
+  auto maybe_input_info = extract_input_lqp_info(node);
+  if (!maybe_input_info) {
+    _apply_to_inputs(node);
+    return;
+  }
+
+  auto input_info = *maybe_input_info;
+
+  // Build a map from parameter ids to their respective expressions.
+  std::map<ParameterID, std::shared_ptr<AbstractExpression>> parameter_mapping;
+  for (size_t parameter_idx = 0; parameter_idx < input_info.subquery_expression.parameter_count(); ++parameter_idx) {
+    const auto& parameter_expression = input_info.subquery_expression.parameter_expression(parameter_idx);
+    parameter_mapping.emplace(input_info.subquery_expression.parameter_ids[parameter_idx], parameter_expression);
+  }
+
+  // Scan for unoptimizable correlated parameter usages, and count correlated predicate nodes.
+  const auto& [not_optimizable, correlated_predicate_node_count] =
+      assess_correlated_parameter_usage(input_info.subquery_expression.lqp, parameter_mapping);
+  if (not_optimizable) {
+    _apply_to_inputs(node);
+    return;
+  }
+
+  // Attempt to pull up all correlated predicate nodes.
+  auto maybe_pull_up_info = attempt_predicate_pull_up(input_info.subquery_expression.lqp, parameter_mapping,
+                                                      correlated_predicate_node_count, false);
+  if (!maybe_pull_up_info) {
+    _apply_to_inputs(node);
+    return;
+  }
+
+  auto pull_up_info = *maybe_pull_up_info;
+  if (input_info.additional_join_predicate) {
```

Comment:
why would there be another join predicate in input_info that is not in pull_up_info? Please doc this.